Lib/test/test_frame.py

```diff
@@ -431,6 +432,116 @@ def test_is_mapping(self):
                 kind = "other"
         self.assertEqual(kind, "mapping")
 
+    def test_proxy_key_stringlikes(self):
+        class StringSubclass(str):
+            pass
+
+        class ImpostorX:
+            def __hash__(self):
+                return hash('x')
+
+            def __eq__(self, other):
+                return other == 'x'
+
+        def f(obj):
+            x = 1
+            proxy = sys._getframe().f_locals
+            proxy[obj] = 2
+            return (
+                proxy.keys() ==  ['obj', 'x', 'proxy'],
+                proxy == {'obj': 'x', 'x': 2, 'proxy': proxy},
+                proxy
+            )
+
+        for obj in StringSubclass('x'), ImpostorX():
+            with self.subTest(cls=type(obj).__name__):
+
+                assertion1, assertion2, proxy = f(obj)
+                self.assertEqual(proxy.keys(),  ['obj', 'x', 'proxy'])
+                self.assertEqual(proxy, {'obj': 'x', 'x': 2, 'proxy': proxy})
+                self.assertTrue(assertion1)
+                self.assertTrue(assertion2)
```

Comment:
Maybe add some message for what those assertions mean (so that on error we directly have a description).